plutus-core/cost-model/budgeting-bench/Benchmarks/CryptoAndHashes.hs

```diff
@@ -1,74 +1,112 @@
+  verification implementations return quickly if some basic precondition isn't
+  satisfied.  For example, at least one Ed25519 signature verification
+  implementation (in cardano-crypto) expects the first three bits of the final
+  byte of the 64-byte signature to all be zero, and fails immediately if this is
+  not the case.  This behaviour isn't documented.  Another example is that ECDSA
+  verification admits two valid signatures (involving a point on a curve and its
+  inverse) for a given message, but in the Bitcoin implementation for
+  EcdsaSecp256k1 signatures (which we use here), only the smaller one (as a
+  bytestring) is accepted.  Again, this fact isn't particularly well advertised.
+  If these basic preconditions are met however, verification time should depend
+  linearly on the length of the message since the whole of the message has to be
+  examined before it can be determined whether a (key, message, signature)
+  triple is valid or not.
+-}
+
+{- | Create a list of valid (key,message,signature) triples.  The DSIGN
+   infrastructure lets us do this in a farily generic way.  However, to sign an
+   EcdsaSecp256k1DSIGN message we can't use a raw bytestring: we have to wrap it
+   up using Crypto.Secp256k1.msg, which checks that the bytestring is the right
+   length.  This means that we have to add a ByteString -> message conversion
+   function as a parameter here.
+-}
+mkBmInputs :: forall v msg .
+    (Signable v msg, DSIGNAlgorithm v, ContextDSIGN v ~ ())
+    => (ByteString -> msg)
+    -> MessageSize
+    -> [(ByteString, ByteString, ByteString)]
+mkBmInputs toMsg msgSize =
+    map mkOneInput (zip seeds messages)
+    where seeds = listOfSizedByteStrings numSamples 128
+          -- ^ Seeds for key generation. For some algorithms the seed has to be
+          -- a certain minimal size and there's a SeedBytesExhausted error if
+          -- it's not big enough; 128 is big enough for everything here though.
+          messages =
+              case msgSize of
+                Arbitrary -> bigByteStrings seedA
+                Fixed n   -> listOfSizedByteStrings numSamples n
+          mkOneInput (seed, msg) =
+              let signKey = genKeyDSIGN @v $ mkSeedFromBytes seed                 -- Signing key (private)
+                  vkBytes = rawSerialiseVerKeyDSIGN $ deriveVerKeyDSIGN signKey   -- Verification key (public)
+                  sigBytes = rawSerialiseSigDSIGN $ signDSIGN () (toMsg msg) signKey
+              in (vkBytes, msg, sigBytes)
 
 benchVerifyEd25519Signature :: Benchmark
 benchVerifyEd25519Signature =
-    createThreeTermBuiltinBenchElementwise name [] pubkeys messages signatures
-           where name = VerifyEd25519Signature
-                 pubkeys    = listOfSizedByteStrings 50 32
-                 messages   = bigByteStrings seedA
-                 signatures = listOfSizedByteStrings 50 64
--- TODO: this seems suspicious.  The benchmark results seem to be pretty much
```

Comment:
This is (finally) explained by the check [here](https://github.com/input-output-hk/cardano-crypto/blob/07397f0e50da97eaa0575d93bee7ac4b2b2576ec/cbits/ed25519/ed25519.c#L103
) that certain bits are zero, which means that if we supply random signatures then seven times out of eight it'll return almost immediately and not do the expensive full verification process (which is what we really want to measure).